### PR TITLE
Removed poco dependency

### PR DESCRIPTION
**Pull request by ahcorde**
As part of the effort to remove Poco dependency described in this issue https://github.com/ros2/ros2/issues/867 this PR makes use of the recent changes in rcutils https://github.com/ros2/rcutils/pull/215 to load, unload and get symbols from shared libraries instead of using Poco.

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

File: rosidl_typesupport_c/CMakeLists.txt [hunk 1]
```diff
@@ -35,23 +31,18 @@ add_library(${PROJECT_NAME}
...
   src/type_support_dispatch.cpp)
-if(Poco_FOUND)
-  target_compile_definitions(${PROJECT_NAME}
-    PRIVATE "ROSIDL_TYPESUPPORT_C_USE_POCO")
-endif()
 if(WIN32)
   target_compile_definitions(${PROJECT_NAME}
     PRIVATE "ROSIDL_TYPESUPPORT_C_BUILDING_DLL")
 endif()
+target_link_libraries(${PROJECT_NAME})
```
Comment (wjwwood) [inline rosidl_typesupport_c/CMakeLists.txt]: What is this for? AFAIK this is a noop.

File: rosidl_typesupport_c/src/type_support_dispatch.hpp [hunk 2]
```diff
@@ -44,15 +41,24 @@ get_typesupport_handle_function(
...
       static_cast<const type_support_map_t *>(handle->data);
     for (size_t i = 0; i < map->size; ++i) {
       if (strcmp(map->typesupport_identifier[i], identifier) != 0) {
         continue;
       }
-      Poco::SharedLibrary * lib = nullptr;
+      rcutils_shared_library_t * lib = nullptr;
+      rcutils_allocator_t allocator = rcutils_get_default_allocator();
+      lib = static_cast<rcutils_shared_library_t *>(allocator.allocate(
+              sizeof(rcutils_shared_library_t), allocator.state));
```
Comment (wjwwood) [inline rosidl_typesupport_c/src/type_support_dispatch.hpp]: This indent looks wrong, it should never be more than two spaces per indent:

```suggestion
        sizeof(rcutils_shared_library_t), allocator.state));
```

Comment (wjwwood) [inline rosidl_typesupport_c/src/type_support_dispatch.hpp]: Should this not be moved down to where `new Poco::SharedLibrary(library_path);` was below?

File: rosidl_typesupport_c/src/type_support_dispatch.hpp [hunk 3]
```diff
@@ -44,15 +41,24 @@ get_typesupport_handle_function(
...
+      if (!lib) {
+        fprintf(stderr, "failed to allocate memory");
```
Comment (wjwwood) [inline rosidl_typesupport_c/src/type_support_dispatch.hpp]: You should throw instead (probably `std::bad_alloc`), because that's what `new` would have done.

File: rosidl_typesupport_c/src/type_support_dispatch.hpp [hunk 4]
```diff
@@ -63,25 +69,26 @@ get_typesupport_handle_function(
           fprintf(stderr, "Failed to find library '%s'\n", library_name);
           return nullptr;
         }
-        lib = new Poco::SharedLibrary(library_path);
+        rcutils_ret_t ret = rcutils_load_shared_library(lib, library_path.c_str());
+        if (ret != RCUTILS_RET_OK) {
+          fprintf(stderr, "Cannot open library %s", library_path.c_str());
```
Comment (wjwwood) [inline rosidl_typesupport_c/src/type_support_dispatch.hpp]: Again, throw, but this time probably `std::runtime_error` or a custom error based on that one.

Comment (wjwwood) [reply rosidl_typesupport_c/src/type_support_dispatch.hpp]: So looking at the rest of the function now, it's really strange that it uses fprintf in some places and throws exceptions in another.

I think we need to clean this up and either throw everywhere or catch all possible exceptions (like when doing `new`) and using fprintf/return nullptr in those cases too. My preference is to throw, since this is obviously a C++ function, as it is in a namespace and is a template function...

Comment (wjwwood) [reply rosidl_typesupport_c/src/type_support_dispatch.hpp]: We should check where it is being called from though, and make sure that place is setup to handle exceptions if/when they occur.

File: rosidl_typesupport_cpp/src/type_support_dispatch.hpp [hunk 5]
```diff
@@ -42,15 +39,24 @@ get_typesupport_handle_function(
...
     const type_support_map_t * map = \
       static_cast<const type_support_map_t *>(handle->data);
     for (size_t i = 0; i < map->size; ++i) {
       if (strcmp(map->typesupport_identifier[i], identifier) != 0) {
         continue;
       }
-      Poco::SharedLibrary * lib = nullptr;
+      rcutils_shared_library_t * lib = nullptr;
+      rcutils_allocator_t allocator = rcutils_get_default_allocator();
+      lib = static_cast<rcutils_shared_library_t *>(allocator.allocate(
```
Comment (wjwwood) [inline rosidl_typesupport_cpp/src/type_support_dispatch.hpp]: Same comments as for the `rosidl_typesupport_c` file.

File: rosidl_typesupport_c/CMakeLists.txt (hunk 1 above)
Comment (ahcorde) [reply rosidl_typesupport_c/CMakeLists.txt]: Removed

File: rosidl_typesupport_c/src/type_support_dispatch.hpp (hunk 2 above)
Comment (ahcorde) [reply rosidl_typesupport_c/src/type_support_dispatch.hpp]: moved bellow

Comment (ahcorde) [reply rosidl_typesupport_c/src/type_support_dispatch.hpp]: fixed

File: rosidl_typesupport_c/src/type_support_dispatch.hpp (hunk 4 above)
Comment (ahcorde) [reply rosidl_typesupport_c/src/type_support_dispatch.hpp]: Throwing `std exceptions` instead of using `fprintf`

File: rosidl_typesupport_cpp/src/type_support_dispatch.hpp (hunk 5 above)
Comment (ahcorde) [reply rosidl_typesupport_cpp/src/type_support_dispatch.hpp]: Fixed in both files

File: rosidl_typesupport_c/CMakeLists.txt [hunk 6]
```diff
@@ -35,22 +30,16 @@ add_library(${PROJECT_NAME}
...
   src/type_support_dispatch.cpp)
-if(Poco_FOUND)
-  target_compile_definitions(${PROJECT_NAME}
-    PRIVATE "ROSIDL_TYPESUPPORT_C_USE_POCO")
-endif()
 if(WIN32)
   target_compile_definitions(${PROJECT_NAME}
     PRIVATE "ROSIDL_TYPESUPPORT_C_BUILDING_DLL")
 endif()
+
```
Comment (dirk-thomas) [inline rosidl_typesupport_c/CMakeLists.txt]: Nitpick: revert unrelated whitespace change.

File: rosidl_typesupport_c/src/type_support_dispatch.hpp [hunk 7]
```diff
@@ -44,44 +43,47 @@ get_typesupport_handle_function(
...
         std::string library_path = rcpputils::find_library_path(library_name);
         if (library_path.empty()) {
-          fprintf(stderr, "Failed to find library '%s'\n", library_name);
-          return nullptr;
+          throw std::runtime_error("Failed to find library '" + std::string(library_name) + "'");
+        }
+
+        try {
+          lib = new rcpputils::SharedLibrary(library_path.c_str());
+        } catch (const std::runtime_error & e) {
```
Comment (dirk-thomas) [inline rosidl_typesupport_c/src/type_support_dispatch.hpp]: This needs to also catch / handle `bad_alloc`.

Same below.

**Comment by ahcorde** [comment]
rcpputils PR is merged, can you review this ? @wjwwood or @dirk-thomas

File: rosidl_typesupport_c/src/type_support_dispatch.hpp [hunk 8]
```diff
@@ -18,15 +18,14 @@
 #include <cstddef>
 #include <cstdio>
 #include <cstring>
+#include <memory>
+#include <stdexcept>
```
Comment (wjwwood) [inline rosidl_typesupport_c/src/type_support_dispatch.hpp:22]: Nitpick: seems like these includes go with the group below not this one.

File: rosidl_typesupport_c/src/type_support_dispatch.hpp [hunk 9]
```diff
@@ -44,44 +43,49 @@ get_typesupport_handle_function(
...
-      lib = const_cast<Poco::SharedLibrary *>(clib);
-      if (!lib->hasSymbol(map->symbol_name[i])) {
-        fprintf(stderr, "Failed to find symbol '%s' in library\n", map->symbol_name[i]);
-        return nullptr;
+      auto clib = static_cast<const rcpputils::SharedLibrary *>(map->data[i]);
+      lib = const_cast<rcpputils::SharedLibrary *>(clib);
+      if (!lib->has_symbol(map->symbol_name[i])) {
+        throw std::runtime_error(
+                "Failed to find symbol '" + std::string(
+                  map->symbol_name[i]) + "' in library");
```
Comment (wjwwood) [inline rosidl_typesupport_c/src/type_support_dispatch.hpp:79]: I think you should prefer to split lines at the + operations rather than in the string call.

File: rosidl_typesupport_c/src/type_support_dispatch.hpp [hunk 10]
```diff
@@ -44,44 +43,49 @@ get_typesupport_handle_function(
...
       if (!map->data[i]) {
         char library_name[1024];
         snprintf(
           library_name, 1023, "%s__%s",
           map->package_name, identifier);
         std::string library_path = rcpputils::find_library_path(library_name);
         if (library_path.empty()) {
-          fprintf(stderr, "Failed to find library '%s'\n", library_name);
-          return nullptr;
+          throw std::runtime_error("Failed to find library '" + std::string(library_name) + "'");
```
Comment (dirk-thomas) [inline rosidl_typesupport_c/src/type_support_dispatch.hpp:62]: Why does this `throw` instead of return `nullptr`?

Same below.

File: rosidl_typesupport_c/src/type_support_dispatch.hpp [hunk 11]
```diff
@@ -63,25 +62,30 @@ get_typesupport_handle_function(
...
           return nullptr;
         }
-        lib = new Poco::SharedLibrary(library_path);
+
+        try {
+          lib = new rcpputils::SharedLibrary(library_path.c_str());
+        } catch (const std::runtime_error & e) {
+          throw std::runtime_error("Could not load library " + std::string(e.what()));
+        } catch (const std::bad_alloc & e) {
+          throw std::bad_alloc();
```
Comment (dirk-thomas) [inline rosidl_typesupport_c/src/type_support_dispatch.hpp:71]: This is basically doing nothing. Please handle the bad alloc case in the same way as the runtime error.

File: rosidl_typesupport_c/src/type_support_dispatch.hpp [hunk 12]
```diff
@@ -63,25 +62,30 @@ get_typesupport_handle_function(
           fprintf(stderr, "Failed to find library '%s'\n", library_name);
           return nullptr;
         }
-        lib = new Poco::SharedLibrary(library_path);
+
+        try {
+          lib = new rcpputils::SharedLibrary(library_path.c_str());
+        } catch (const std::runtime_error & e) {
+          throw std::runtime_error("Could not load library " + std::string(e.what()));
```
Comment (dirk-thomas) [inline rosidl_typesupport_c/src/type_support_dispatch.hpp:69]: The exception message doesn't include the library path for which the error occurred and is therefore not really helpful for users. Arguably the lower level API in `rcpputils` / `rcutils` should probably already include the library path in the exception / error message.

File: rosidl_typesupport_c/src/type_support_dispatch.hpp [hunk 13]
```diff
@@ -63,25 +62,34 @@ get_typesupport_handle_function(
           fprintf(stderr, "Failed to find library '%s'\n", library_name);
           return nullptr;
         }
-        lib = new Poco::SharedLibrary(library_path);
+
+        try {
+          lib = new rcpputils::SharedLibrary(library_path.c_str());
+        } catch (const std::runtime_error & e) {
+          throw std::runtime_error(
+                  "Could not load library " + library_path + " :" +
```
Comment (dirk-thomas) [inline rosidl_typesupport_c/src/type_support_dispatch.hpp:70]: Nitpick: `": "` instead of `" :"`.

Same below.

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/9928/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5538/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8076/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/9774/)

**Comment by ahcorde** [comment]
Warning on Windows appears on spdlog_vendor which is strange because the code has not been modified for 15 days

```
14:24:43   CMake Warning (dev) at C:/Program Files/CMake/share/cmake-3.17/Modules/GNUInstallDirs.cmake:225 (message):
14:24:43     Unable to determine default CMAKE_INSTALL_LIBDIR directory because no
14:24:43     target architecture is known.  Please enable at least one language before
14:24:43     including GNUInstallDirs.
14:24:43   Call Stack (most recent call first):
14:24:43     CMakeLists.txt:9 (include)
14:24:43   This warning is for project developers.  Use -Wno-dev to suppress it.
```
